### ISSUE TITLE
your thought process in plain english

### ISSUE THREAD

**Issue by krompir2** [typestack/class-transformer#109]
Hey man! This project is a huge time saver and I am happily using it in Angular apps.

Unfortunately I have some issue, not with the code performing unexpectedly, but with understanding the code. It would be a great learning experience if you could add a few lines in plain english, explaining your thought process when writing this code. Kinda like the big picture of this library. 

As a self educated developer, I am always interested in how other, better developers think and what made them think that way, so I can learn from them. 

I hope this post makes sense and I am looking forward to hear from you!
Regards, Mitt

**Comment by gempain** [comment]
Hello,
I will agree with you on that one. Working on pull requests can be hard sometimes when you don't actually have enough comments in the code.
However, I find most of the code self explanatory because I feel it was written well, even though certain parts would need a bit of refactoring or more comments, especially the serialising / deserialising part.
Feel free to make a PR if you want to improve some areas of the code or add comments here and there.

**Comment by krompir2** [comment]
It's not that the code is badly written, it's that I am bad at reading it.  For example, if I'd be given a task to write a utility tool to transform object tree into json and vice versa, I'd have no idea where to even start, because I am missing some crucial logic to perform such task. It's much harder to understand code, specially if you are a few nested methods deep, if you don't have the big picture in mind. Comments help tremendously, but not with the big picture, because they are scrambled all over the place.

**Comment by NoNameProvided** [comment]
Actually, there is a lot of opportunities to refactor the code as well. I totally feel your problem, there are parts which are hard to understand for me as well every time I read it. We will improve in this.

**Comment by NoNameProvided** [comment]
I am closing this, as "solved". Once I get to refactor all the parts for the lib, the code will be properly documented. I am also in the process of creating proper documentation for the TypeStack projects (currently working on documentation for TypeDI.)
